platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,852 @@
+      - This is because each page is a separate Formik form, so the `Router` is
+        responsible for stitching all these individual forms together into one
+        cohesive experience
+1. Build a list of routes for the form flow from `Chapter` and `Page`
+   components' `path` props
+    - **Important:** `Route`s define valid routes in a form, but those pages
+      live "outside" the normal form flow; these are used for pages such as
+      `/form-saved` and `/error`
+    - `Page`s nested in `Chapter`s will have their paths prefixed with the
+      `Chapter` path
+    - For example:
+      ```jsx
+      <Chapter path="/personal-information">
+        <Page path="/name">...</Page>
+      </Chapter>
+      ```
+      This will result in a path definition of `/personal-information/name`
+1. Re-compute the list of routes before each page navigation (below)
+    - This ensures any data changes that might affect the routes are captured
+1. Provide callbacks in the React context to navigate forward and backward
+   through the list of computed routes
+
+##### `Chapter` Component
+**Purpose:**
+The **optional** `Chapter` component will be used to render the VA.gov "chrome"
+for multi-page forms which have multiple chapters. This "chrome" includes the
+form title, current chapter name, and progress bar.
+
+**Usage requirements:**
+- `Chapter` may only be a child of `Router`
+- Its only children may be `Page`s
+- A `Chapter` must have children
+
+**Responsibilities:**
+- Render the chapter title and progress bar
+- Render the child `Page`s
+
+##### `Page` Component
+**Purpose:**
+A `Page` is used to define the path for a page and render its form contents and
+the normal page chrome.
+
+**Usage requirements:**
+A `Page` may be the child of either `Router` directly or `Chapter`.
+
+**Responsibilities:**
+- If no `overrideFormik` prop is passed
+  - Render a [`<Formik>`](https://formik.org/docs/api/formik) component and pass
+    it some default props for `initialValues`, `validate`, and `onSubmit`
+  - If the save-in-progress React context is available, `Page` will use it to
+    auto-save the form
+- Render the children, passing them the extra props for form state
+  - For an example usage, see [below](#functionality-array-pages)
+  - **Caveat:** This means when the children of `Page` are a collection of
+    `<Field>`s (or similar) instead of the functional component in the example
+    below, they'll all have "extra" props for the form context
+    - This is the simplest way to support both simple and complex form pages,
+      but may be subject to change later if we can figure out a better way
+- Render the form navigation buttons "Back" and "Continue"
+- Prevent navigation if there are validation errors
+
+##### `Route` Component
+**Purpose:**
+The `Route` will be re-exported from `react-router-dom`. The purpose of
+re-exporting is twofold:
+- Simplifies imports
+- Ensures only one version of `react-router-dom` is used
```

Comment:
What is going to keep someone from importing from `react-router-dom`, since it would be in `node_modules` as a dependency of `formulate`, if I'm understanding this correctly.

Reply:
Nothing. :man_shrugging: But if they import the component from the same dependency as Formulate, it'll be the same version, so that's fine.

The problem I foresee is having a peer dependency of `react-router-dom` that doesn't match the version Formulate uses. So if we try to use the `<Route>` from the `react-router-dom@1.0.0` specified in the application's `package.json` in the `<Router>` from Formulate, which uses the `Router` from `react-router-dom@2.0.0`, we may have a problem.

This library won't attempt to prevent engineers from doing anything. It'll attempt to provide engineers with the tools they need to do what they need the right way. In this case, it means ensuring the `Route` is using the same version as the `Router`.

Reply:
So this re-exporting doesn't necessarily ensure that only one version of `react-router-dom` is used? I don't think we need to try to make that a thing, but I was curious about the statement itself given some past experiments.

There may be an assumption here that's interfering with things. For the build of `formulate` that gets published, will it already have the relevant stuff from `react-router-fom` in the bundle? When I made the original comment I was thinking that that wouldn't be the case.

Reply:
Assuming tree shaking is working properly in the application using Formulate, if it doesn't import `Router` or any of the routing components, the application bundle will not include `react-router-dom` because of Formulate. When an application imports `Router` or any of the routing components, the application bundle will include (parts of? :thinking:) `react-router-dom` as a dependency of the Formulate package.